### DISCUSSION TITLE
typescript adapter methods not compatible with mikroorm populate

### DISCUSSION THREAD

**Discussion by corentingosselin** [mikro-orm/mikro-orm#4953] [General]
### Describe the bug

This is the error:

```
Entity 'SchoolEntity' does not have property 'campuses'

ValidationError: Entity 'SchoolEntity' does not have property 'campuses'
    at Function.invalidPropertyName (/Users/prodigy/Documents/campuscalendar/node_modules/.pnpm/@mikro-orm+core@5.7.12_@mikro-orm+migrations@5.7.12_@mikro-orm+mysql@5.7.12/node_modules/@mikro-orm/core/errors.js:57:16)
    at SqlEntityManager.preparePopulate (/Users/prodigy/Documents/campuscalendar/node_modules/.pnpm/@mikro-orm+core@5.7.12_@mikro-orm+migrations@5.7.12_@mikro-orm+mysql@5.7.12/node_modules/@mikro-orm/core/EntityManager.js:1228:44)
    at SqlEntityManager.find (/Users/prodigy/Documents/campuscalendar/node_modules/.pnpm/@mikro-orm+core@5.7.12_@mikro-orm+migrations@5.7.12_@mikro-orm+mysql@5.7.12/node_modules/@mikro-orm/core/EntityManager.js:100:31)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
``` 

My populate code:

```
    await this.orm.em.getRepository(SchoolEntity).findAll({
      populate: ['campuses', 'classYears'],
    });
``` 

I have a simple SchoolEntity, with methods adapter.
I'm using this to convert collection into array. This is necessary because I have High level interface managing response, dto, entity. We are inside a monorepo with angular application, with shared api interfaces. So we cannot import mikroorm collection.

```
import { Entity, Property, OneToMany, Collection } from '@mikro-orm/core';
import { ClassYearEntity } from './class-year.entity';
import { CampusEntity } from './campus.entity';
import { BaseEntity, Campus, ClassYear, School, SchoolResponse } from '@campuscalendar/shared/api-interfaces';

@Entity()
export class SchoolEntity extends BaseEntity implements School {

  @Property()
  name!: string;

  @OneToMany(() => CampusEntity, campus => campus.school)
  private _campuses = new Collection<CampusEntity>(this);

  @OneToMany(() => ClassYearEntity, classYear => classYear.school)
  private _classYears = new Collection<ClassYearEntity>(this);

  // Adapter methods
  get campuses(): Campus[] {
    return this._campuses.toArray();
  }

  set campuses(campuses: CampusEntity[]) {
    this._campuses.set(campuses);
  }

  get classYears(): ClassYear[] {
    return this._classYears.toArray();
  }

  set classYears(classYears: ClassYearEntity[]) {
    this._classYears.set(classYears);
  }
}

export interface School extends Entity {
    name: string;
    campuses: Campus[];
    classYears: ClassYear[];
}
```    

```
import { BaseEntity, Campus } from '@campuscalendar/shared/api-interfaces';
import { Entity, ManyToOne, Property } from '@mikro-orm/core';
import { SchoolEntity } from './school.entity';

@Entity()
export class CampusEntity extends BaseEntity implements Campus {

  @Property()
  name!: string;

  @ManyToOne(() => SchoolEntity)
  school!: SchoolEntity;

}

export interface Campus extends Entity {
    name: string;   
}
```     

To fix this, I have to populate manually:

```
  @UseRequestContext()
  async getSchool() {
    const result = await this.orm.em.getRepository(SchoolEntity).findAll();
    const school = result[0];

    const campuses = await this.orm.em.find(CampusEntity, { school: school });
    school.campuses = campuses;

    const classYears = await this.orm.em.find(ClassYearEntity, {
      school: school,
    });
    school.classYears = classYears;
    return school.toResponse();
  }

  //Inside school entity:

  toResponse(): SchoolResponse {
    return {
      id: this.id,
      created_at: this.created_at,
      updated_at: this.updated_at,
      name: this.name,
      campuses: this._campuses.toArray(),
      classYears: this._classYears.toArray()
    };
  }
```    

This solution is not the best, and I can't build the response I want with sub entities using this toResponse function.

### Reproduction

Just use custom function adapter, mikro orm does not seems to detect the field.

### What driver are you using?

@mikro-orm/mysql

### MikroORM version

5.9.4

### Node.js version

v18.17.0

### Operating system

macos

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
You cannot populate things that are not declared ORM properties - those marked with the decorators. Getters are runtime construct, the ORM does not know what your getter implementation is doing.

Not sure why do you want to do this in the first place, `Collection` instance is array-like interface, the place where this gets converted to array is serialization. Instead of converting collections like this, you should convert the whole entity when you need to work with the DTO (via `wrap(e).toObject()`).

**Reply by corentingosselin** [reply]
Thank you :) I will try to implement this
